### PR TITLE
Issue2905 Steam District Heating System

### LINKED ISSUE #2905: Steam District Heating - Complete System
@bravache and @JayHuLBL: Please do not create the branch for this issue until PRs for Issues #2701 and #2904 have (at a minimum) been merged. A PR for Issue #2848 will be opened ASAP if @SaranyaAnbarasu and I can complete the model improvements requested by Michael in time; however, it is not required. With that said, I wanted to open this "system-level" issue prematurely so that you are aware of this final requirement for the OSMSES Conference paper. This is the last time-sensitive issue for the steam district heating models. These models need to be merged to `lbl-srg/issue2905_steamsystem` before the paper submission deadline of **Thursday, March 10, 2022 (time zone CET, GMT+1, Aachen, Germany)**. (Our **internal deadline** should be **end-of-business PT on Wed. March 9**.)

This issue is to to add the subsystem (Steam Plant) and system-level example (complete steam district heating system) to the `Experimental.DHC` package. In addition to the complete system example (`Steam` model to be located in `DHC.Examples`), the below subsystem model will be contributed (highlighted in green). This work is part of the URBANopt project.

image

This subsystem and system models use components from previously opened issues (mentioned above). Once the component branches have been merged from fork to lbl-srg, they need to be pulled into this system-level branch before the new developments can be added. Here is a graphical representation of the current status and requirements:

[Updated 3/2/22 with dependency with Network models]
[Updated 3/4/22 with PRV PR]
[Updated 3/9/22 with Networks PR]
[Updated 4/4/22 with components/subsystems merged with development branches]
[Updated 4/12/22 PR for complete system opened.]
image
I am fully available to work on these developments until the March 10th deadline. If you have any questions or if I can help in any way, please let me know. My sincerest thanks!

### PR DESCRIPTION
**Pull request by bravache**
This closes #2905

### CODE REVIEW

**Comment by bravache** [comment]
@khinkelman 

TL;DR : Please make sure that the following examples run without any unspecified initial condition warnings:
- Buildings.Experimental.DHC.Plants.Steam.Examples.SingleBoiler
- Buildings.Experimental.DHC.EnergyTransferStations.BaseClasses.Validation.CollectorDistributor

---

Hi Katy, I've reverted your change to `Buildings.Experimental.DHC.Plants.BaseClasses.PartialPlant` as it was breaking the original chiller (`Buildings.Experimental.DHC.Plants.Cooling.ElectricChillerParallel`). This was because even disabled, the parameter `fue` should have a value for the model to be initialized (the `enable` option is only a visual indicator for users in the dialog box, but all parameters should have a value). This is not an issue if it is an array with size 0 (the way it was previously).
I've reverted to that behavior and also changed your boiler (`Buildings.Experimental.DHC.Plants.Steam.SingleBoiler`) and introduce a parameter `fueBoi` to hide the unnecessary array to a user of that class. Future plants might have more than one fuel.

The following example failed because of unspecified initial conditions (this is just a warning in Dymola, but the CI test will return an error):
- Buildings.Experimental.DHC.Plants.Steam.Examples.SingleBoiler
- Buildings.Experimental.DHC.Loads.Steam.BaseClasses.Examples.SteamTrap
- Buildings.Experimental.DHC.Networks.Steam.Examples.DistributionCondensatePipe
- Buildings.Experimental.DHC.Examples.Steam.SingleBoiler

I've fixed some of those issues (the steam trap and the distribution condensate pipe should now run without warnings) but there is something else going on with the single boiler that I can't pinpoint. Can you take a look and see if that's something you can solve.

Also, I believe that the refactor of `PartialConnection2Pipe` and `PartialDistribution` had some unintended consequences, and broke the `CollectorDistributor`, along with all of the examples that use it. For troubleshooting, you can focus on `Buildings.Experimental.DHC.EnergyTransferStations.BaseClasses.Validation.CollectorDistributor` which is the minimal viable use of that class. Other examples that uses that collector are in `Buildings.Experimental.DHC.EnergyTransferStations.Combined.Examples`.

**Comment by khinkelman** [comment]
Thanks @bravache, we will work on these ASAP.

**Comment by khinkelman** [comment]
To do:
- [ ] Initial conditions for `Buildings.Experimental.DHC.Plants.Steam.Examples.SingleBoiler`
- [ ] Initial conditions for `Buildings.Experimental.DHC.Examples.Steam.SingleBoiler`
- [ ] Diagnose `CollectorDistributor` errors via `Buildings.Experimental.DHC.EnergyTransferStations.BaseClasses.Validation.CollectorDistributor`
- [ ] Verify CI test errors resolved
- [ ] Rerun unit tests

**Comment by bravache** [comment]
@dhblum Hi David,

This branch contains the Steam District Heating components and system that was developed by Katy and Saranya. I have reviewed all of the pieces individually and combined, and it looks good to me. I was hoping that you could give it another look to see if I might have missed anything (and it would also be a good exercise to get into the UrbanOpt project).

There is more information about the different pieces on the related Pull Requests and issues that led to this branch:
Steam Valves: https://github.com/lbl-srg/modelica-buildings/pull/2908
Steam Boiler: https://github.com/lbl-srg/modelica-buildings/pull/2701
Steam Network: https://github.com/lbl-srg/modelica-buildings/pull/2914

Please have a look and if everything looks ok, this is ready to be merged. 

@khinkelman It looks like there are issues with openmodelica and OCT for running some of the modified/new examples, which would need to be addressed. I don't know if your team is equipped with reproducing those errors, but if not, I believe that David (or someone else at LBL) might be able to track them down.

**Comment by dhblum** [comment]
@bravache Ok thanks, sure I'll take a look, also at the OM and OCT errors.

@khinkelman Nice to meet you!

**Comment by dhblum** [comment]
@khinkelman Just an update here, I was able to reproduce the Optimica unit test error on my machine.  It is:

```
Error in flattened model:
  Cannot differentiate call to function without derivative or smooth order annotation 'Modelica.Media.Water.IF97_Utilities.BaseIF97.Basic.g2(pla.boi.vol.p, pla.boi.heaCapDry.T, true)' in equation:
   (Modelica.Media.Common.GibbsDerivs(, , , temp_00000037, , , temp_00000040, , , , )) = Modelica.Media.Water.IF97_Utilities.BaseIF97.Basic.g2(pla.boi.vol.p, pla.boi.heaCapDry.T, true)
```
Media models in MSL compile ok and other steam example models compile ok (``Buildings.Experimental.DHC.Plants.Steam.BaseClasses.Examples.BoilerPolynomial`` and ``Buildings.Experimental.DHC.Plants.Steam.BaseClasses.Examples.ControlVolumeEvaporation``).  The problem seems to be isolated to ``Buildings.Experimental.DHC.Plants.Steam.SingleBoiler``, as I've created a version of ``Buildings.Experimental.DHC.Plants.Steam.Examples.SingleBoiler`` with only the plant, resistance, and condensing volume components that compiles ok in Dymola (seemingly the minimum components around the plant model needed).  I further removed all components from ``Buildings.Experimental.DHC.Plants.Steam.Examples.SingleBoiler`` except the feedwater pump and boiler (each with just constant 0 control inputs), and still get the same error.  But, if I simply add a pump to ``Buildings.Experimental.DHC.Plants.Steam.BaseClasses.Examples.BoilerPolynomial``, that model compiles ok.  

So, maybe it has to do with the ``PartialPlant`` base class the single boiler plant extends, or how it extends it.  Need to do more testing.

**Comment by dhblum** [comment]
An update on Optimica compilation errors.  A simple test shows that the placement of the expansion tank makes all the difference:

The following compiles OK:
ok
While the following does NOT compile OK:
error

Therefore, taking the failing example model in question (``Buildings.Experimental.DHC.Plants.Steam.Examples.SingleBoiler``) and moving the expansion tank to outside of the plant model for demonstration purposes, the following compiles OK:
ok1

While the following does NOT compile OK (this is representative of the current configuration in the failing model):
error1

File: Buildings/Experimental/DHC/BaseClasses/Steam/PartialSaturatedControlVolume.mo [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+  parameter Modelica.Units.SI.Volume V "Total volume";
+  // Initialization
+  parameter Modelica.Units.SI.Volume VWat_start=V/2
+    "Start value of liquid volume"
+    annotation (Dialog(tab="Initialization"));
+
+  // Variables
+  MediumWat.ThermodynamicState stateWat(p=p, T=T)
+    "Saturated state, liquid water";
+  MediumWat.ThermodynamicState stateSte(p=p, T=T)
```
Comment (dhblum) [inline Buildings/Experimental/DHC/BaseClasses/Steam/PartialSaturatedControlVolume.mo:26]: @khinkelman Can you confirm if this should actually be ``MediumSte`` instead of ``MediumWat``?  Making the change helps resolve one of the OpenModelica compilation errors.

**Comment by dhblum** [comment]
@SaranyaAnbarasu After merging the latest master and re-running unit tests, I found an issue where I think the ``V_flow`` is specified incorrectly in the performance data of the pumps in the examples.  Specifically, the density is multiplied by mass flow, where it should be divided.  Please see this commit for my corrections: https://github.com/lbl-srg/modelica-buildings/pull/2961/commits/7873de919ce96725a20da9feb64b51d8bec2eb03.

This resulted in more reasonable results in the unit tests for pump powers given a new model for mover efficiency that has been implemented.  However, the example ``Buildings.Experimental.DHC.Examples.Steam.SingleBoiler`` now shows that the feedwater pump control gets saturated, with the measured and set point values off by a little bit.  Before, the control signal was constant at about 0.71. See the new plot here:

Screen Shot 2023-05-10 at 11 01 31 AM

Do you mind taking a look and just letting me know if this behavior is ok, or if another change should be made based on the example model intent?  I can then update the unit test reference results accordingly.  Thank you!

**Comment by dhblum** [comment]
@SaranyaAnbarasu Another update, with some potentially good news: The Buildings Library is being tested with an updated version of OpenModelica (1.22-dev).  This new version has made some improvements since the last.  I tested four example steam models with the new version locally and they finished the simulations successfully.  I took a brief look at some results and they looked reasonable.  I'm going to see if they pass the unit tests with OpenModelica before adding them to the exclude list.

**Comment by dhblum** [comment]
Updated sizing of feedwater pump in ``Buildings.Experimental.DHC.Examples.Steam.SingleBoiler`` and corresponding reference results.

``Buildings.Experimental.DHC.Examples.Steam.SingleBoiler`` simulates in OM 1.22.0~dev-161-g8eff86a on Ubuntu 20.

``Buildings.Experimental.DHC.Plants.Steam.BaseClasses.Examples.ControlVolumeEvaporation`` fails to simulate with OM (same version as above), and terminates with:
```
IF97 medium function tsat called with too low pressure p = 0 Pa <= 611.657 Pa (triple point pressure)
simulation terminated by an assertion at initialization
Simulation process failed. Exited with code 255.
```

Will let travis unit tests run and see if there are other failures which need to be addressed besides this one.

**Comment by dhblum** [comment]
After touching base with @SaranyaAnbarasu I've fixed all issues I can and added some failing unit tests to exclude list.  Unit tests should pass and @mwetter I believe this is ready for a final review, which I would appreciate if you can take.  I know there is still one last todo besides your review:

- [ ] Update release notes

File: Buildings/Experimental/DHC/BaseClasses/Steam/PartialSaturatedControlVolume.mo [hunk 2]
```diff
@@ -0,0 +1,298 @@
+within Buildings.Experimental.DHC.BaseClasses.Steam;
+partial model PartialSaturatedControlVolume
+  "Partial control volume for evaporation/condensation processes"
+  extends Buildings.BaseClasses.BaseIcon;
+  extends Buildings.Experimental.DHC.BaseClasses.Steam.PartialTwoPortTwoMedium(
+    p_start=1e6,
```
[the remark this answers is not in the document]
Comment (dhblum) [reply Buildings/Experimental/DHC/BaseClasses/Steam/PartialSaturatedControlVolume.mo:6]: It can be changed to Medium.p_default so inherit from medium model.

**Review by mwetter — APPROVED** [review]
